internal/interceptors.go

```diff
@@ -0,0 +1,181 @@
+// Copyright (c) 2020 Temporal Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package internal
+
+import (
+	"github.com/uber-go/tally"
+	"time"
+)
+import "go.uber.org/zap"
+
+// WorkflowInterceptorFactory is used to create a single link in the interceptor chain
+type WorkflowInterceptorFactory interface {
+	// NewInterceptor creates an interceptor instance. The created instance must delegate every call to
+	// the next parameter for workflow code function correctly.
+	NewInterceptor(info *WorkflowInfo, next WorkflowInterceptor) WorkflowInterceptor
```

Comment:
Should we pass workflow context as parameter?

Reply:
This code doesn't run in the context of the workflow thread. So I don't want to give the context here.